File: src/main/java/io/xol/chunkstories/api/events/voxel/VoxelModificationEvent.java
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.api.events.voxel;

import io.xol.chunkstories.api.events.EventListeners;
import io.xol.chunkstories.api.world.cell.CellData;

public class VoxelModificationEvent extends VoxelEvent {
	// Every event class has to have this

	static EventListeners listeners = new EventListeners(VoxelModificationEvent.class);

	@Override
	public EventListeners getListeners() {
		return listeners;
	}

	public static EventListeners getListenersStatic() {
		return listeners;
	}

	// Specific event code

	final WorldModificationCause modificationCause;
	CellData newData;

	public VoxelModificationEvent(CellData context, CellData newData, WorldModificationCause cause) {
		super(context);
		this.newData = newData;
		this.modificationCause = cause;
	}

	public CellData getNewData() {
		return newData;
	}

	public WorldModificationCause getModificationCause() {
		return modificationCause;
	}

	public ModifiationType getModification() {
		if (getContext().getVoxel().isAir())
			return ModifiationType.PLACEMENT;
		else {
			if (newData.getVoxel().isAir())
				return ModifiationType.DESTRUCTION;
			else
				return ModifiationType.REPLACEMENT;
		}
	}

	public enum ModifiationType
	{
		DESTRUCTION, PLACEMENT, REPLACEMENT;
	}
}

File: src/main/java/xyz/chunkstories/api/events/voxel/VoxelModificationEvent.kt
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.api.events.voxel

import xyz.chunkstories.api.events.EventListeners
import xyz.chunkstories.api.world.cell.Cell

open class VoxelModificationEvent(context: Cell, newData: Cell, val modificationCause: WorldModificationCause) : VoxelEvent(context) {

    override val listeners: EventListeners
        get() = listenersStatic
    var newData: Cell
        internal set

    val modification: ModifiationType
        get() = if (context.voxel.isAir())
            ModifiationType.PLACEMENT
        else {
            if (newData.voxel.isAir())
                ModifiationType.DESTRUCTION
            else
                ModifiationType.REPLACEMENT
        }

    init {
        this.newData = newData
    }

    enum class ModifiationType {
        DESTRUCTION, PLACEMENT, REPLACEMENT
    }

    companion object {
        // Every event class has to have this

        var listenersStatic = EventListeners(VoxelModificationEvent::class.java)
            internal set
    }
}
